### PR TITLE
fix some macro expansion scope bugs

### PR DESCRIPTION
**Pull request by vtjnash**
This fixes macro hygiene for cases where a macro returns a call to another macro. It's actually a bit like doing the inverse of #10940.

### CODE REVIEW

File: test/workspace.jl [hunk 1]
```diff
@@ -48,18 +48,23 @@ mktempdir() do dir
...
                 using Test22101
-                f22101()
-                workspace()
-                using Test22101
-            finally
-                splice!(LOAD_PATH, 1)
-                splice!(Base.LOAD_CACHE_PATH, 1)
+                @test isdefined(Main, :f22101)
+                @test f22101()::Vector{Int} == collect(1:10)
+                @eval workspace() using Test22101
```
Comment (tkelman) [inline test/workspace.jl]: this looks strange, what is this doing?

Comment (vtjnash) [inline test/workspace.jl]: running eval in a new workspace

Comment (tkelman) [inline test/workspace.jl]: is that any different than
```
workspace()
@eval Main using Test22101
```
?

Comment (vtjnash) [inline test/workspace.jl]: `Core.Main`, and no, those would name the same module

Comment (tkelman) [inline test/workspace.jl]: Is this the only reason you made `workspace()` return the module or is there some other motivating reason for that?

Comment (vtjnash) [inline test/workspace.jl]: Yes, I just found it awkward that there was no particularly easy or correct way to reference the new `workspace()`. `Core.Main` is supposed to basically work, but probably isn't entirely correct.

Comment (tkelman) [inline test/workspace.jl]: > probably isn't entirely correct.

How so? It does `m = Core.Main` then `return m` a few lines later

Comment (vtjnash) [inline test/workspace.jl]: True, still doesn't make it correct. Although `workspace()` itself isn't really "correct", so it's a rather weak statement anyways.

Comment (tkelman) [inline test/workspace.jl]: incorrect how?

Comment (vtjnash) [inline test/workspace.jl]: What does it do? We know that it sort of does something, most of the time – but that's not very helpful for making it "correct".

Comment (tkelman) [inline test/workspace.jl]: what the docstring says it should do?

still not seeing the motivation for changing its behavior to return a module. that's a separate change from "fix some macro expansion scope bugs" and not discussed anywhere in any linked issues that I can find

Comment (vtjnash) [inline test/workspace.jl]: This is a scope test, and currently printing a warning that it's not working correctly. And I'm changing the scope resolution algorithm. So I needed to make sure that this test is working first.

Comment (tkelman) [inline test/workspace.jl]: is this equivalent to
```
workspace()
@eval Core.Main using Test22101
```
or not? If not, exactly what's different? If so, then you shouldn't need to change the behavior of workspace to test this

Comment (vtjnash) [inline test/workspace.jl]: If the test passes, they were the same. If not, it may or may not be a bug.

File: test/docs.jl [hunk 2]
```diff
@@ -645,8 +645,8 @@ f12593_2() = 1
 @test (Docs.@repl 0) !== nothing
 
 let t = @doc(DocsTest.t(::Int, ::Int))
-    @test docstrings_equal(Docs.@repl(DocsTest.t(0, 0)), t)
-    @test docstrings_equal(Docs.@repl(DocsTest.t(::Int, ::Int)), t)
+    @test docstrings_equal(Docs.@repl((@__MODULE__).DocsTest.t(0, 0)), t)
+    @test docstrings_equal(Docs.@repl((@__MODULE__).DocsTest.t(::Int, ::Int)), t)
```
Comment (JeffBezanson) [inline test/docs.jl]: Is this change actually necessary? I'm worried that resolution of identifiers in macro arguments could depend on the number of layers of macros involved.

**Comment by JeffBezanson** [comment]
I generally think #10940 is a good idea, and we should do it. Could we just revive that instead? Since I like #10940, I'm worried that I might not like the inverse of it :)

**Comment by vtjnash** [comment]
It's the inverse, but not actually the reverse (or something like that. yeah, it's confusing). IIUC, with this fix, implementing that PR simply means that we need to automatically wrap all arguments to macros in `Expr(:esc)` (but rename it `Expr(:hygienic)` and only apply it to symbols).

File: test/docs.jl (hunk 2 above)
Comment (JeffBezanson) [inline test/docs.jl]: The behavior I think we want is for `DocsTest` to resolve to the call environment no matter what the macros do (unless they very specifically mangle the identifier to something else).

**Comment by JeffBezanson** [comment]
If I remember, one thing I liked about #10940 is that it basically leaves macro arguments as they are now, and instead wraps symbols inside quoted code in macros with `Expr(:hygienic)`. That strikes me as more user-friendly.

**Comment by vtjnash** [comment]
> is that it basically leaves macro arguments as they are now

If my recollection of my rebase attempt serves though, it actually couldn't leave them alone. It sometimes mangled them into `GlobalRef`, unless it guesses that resolve-scopes would have eventually made it a local symbol under the assumption that any further macro expansion will use the symbol as a local variable in the current scope and won't introduce any new scopes (e.g. basically  the same assumptions that the code makes now).

Calling this "the inverse" was perhaps too tongue-in-cheek, since I meant it was approaching the fix to a particular bug from the other direction (a bug that the implementation in #10940 sadly doesn't actually end up fixing completely, I think it just reduced it in severity).

But anyways, wrapping quoting / interpolation instead of wrapping the arguments is another possible avenue to implementing #10940 that becomes much easier and more correct with this PR. In this PR, I called that operation `Expr(:hygienic-scope)`. The requisite lowering changes would be to make `@hygienic quote $expr end` (or equivalently `quote $expr end` when written inside a macro body) to produce:
```Expr(:hygienic-scope, Expr(:quote, Expr(:begin, Expr(:esc, expr))))```
(the macro itself then would not imply any hygienic-scope operation – the return value would just be used directly).
I agree, that does sound like the more user-friendly approach.

File: test/docs.jl (hunk 2 above)
Comment (JeffBezanson) [inline test/docs.jl]: Edit: this turns out to be needed because `@repl` explicitly looks up everything in `Main`; that confused me.

**Review by JeffBezanson — APPROVED** [review]

File: test/workspace.jl (hunk 1 above)
Comment (tkelman) [inline test/workspace.jl]: How could it not pass if it is returning `Core.Main` ? This change seems unnecessary and unrelated.

Comment (vtjnash) [inline test/workspace.jl]: As I said, if it passes, they might be the equivalent. If not, it might be a bug.

Comment (tkelman) [inline test/workspace.jl]: You've made it return Core.Main - if functions suddenly returned something other than what they say they do, we'd have bigger problems?

Comment (vtjnash) [inline test/workspace.jl]: Perhaps. But note that these expressions are run in different scopes.

**Comment by StefanKarpinski** [comment]
This change is breaking everything and needs to be reverted or fixed ASAP.
